### PR TITLE
Fix recovery from coin db crashes (and dbcrash.py test)

### LINKED ISSUE #3439: dbcrash.py failure
#### If so, please list the steps to reproduce below:
1. git checkout develop && git pull
2. make
3. `./test/functional/dbcrash.py`

### What version of Dash Core are you using?
commit: de931a25a3fa7ac76aaad17de4517224dbf90de7

### PR DESCRIPTION
**Pull request by 10xcryptodev**
:beetle: improve evodb consistency recovering from dbcrash

This PR fix #3439

### CODE REVIEW

**Comment by UdjinM6** [comment]
Thanks for finding the root of the issue! 👍This fixes `dbrcrash.py` indeed however it "over-fixes" the initial issue imo - we should only commit to evodb when the cache is flushed to ensure consistency. Also, changes in `DisconnectBlock` should be done the other way around - should update the cache and evodb at the very end of the function to ensure additional indexes were also updated successfully before that. Pls see 441069949cfbbfce206c47f7b81d46fd0098ba7f

**Comment by 10xcryptodev** [comment]
@UdjinM6 thanks for reviewing, i only kept this line https://github.com/dashpay/dash/commit/441069949cfbbfce206c47f7b81d46fd0098ba7f#diff-24efdb00bfbe56b140fb006b562cc70bL4360
all the other changes are ok

**Comment by UdjinM6** [comment]
Hmmm.. no, keeping this line would simply let the node to ignore the inconsistency. So if it's crashing without it, the real issues is somewhere else :( The patch (without this line) would still be a step forward though but not a complete fix.

**Comment by 10xcryptodev** [comment]
@UdjinM6 this line is just moving back to the last state when a crash happens in the middle of db writing. txdb has DB_BEST_BLOCK and DB_HEAD_BLOCKS as you can see here https://github.com/dashpay/dash/blob/develop/src/txdb.cpp#L111 and the dbcrash simulation makes DB_BEST_BLOCK and DB_HEAD_BLOCKS point to different hashes. Without this line will be necessary to implement something similar in evodb.
I don't see a simple solution without this line, but we can remove and investigate this in the future

**Comment by UdjinM6** [comment]
No, this line just "fakes" it and makes it look like evodb is at the right point when it is not. The actual fix would be to flush evodb in one more place https://github.com/dashpay/dash/commit/13afebb4d3fbf455be63a43017c334ad4b38e0fb#diff-c865a8939105e6350a50af02766291b7R1903-R1908. But then again, this only ensures consistency of block hashes in coin cache and in evodb are pointing to, which "fixes" `dbcrash.py` but the evodb itself and additional indexes are going to be left in a weird state after `ReplayBlocks` in case of a fork. This is because `DisconnectBlock` reverts some evodb/indexes related data but the new data is not applied in `RollforwardBlock`. It's not that trivial to replicate `ConnectBlock` logic to do this without copy-pasting it seems (which would be a bad idea IMO) and directly using `ConnectBlock` is not quite possible here either. I think it's better to simply stop supporting this case and instead require to fully reindex in case of a crash during a fork e.g. acc46000ac. Forks are reeeeally rare these days anyway so this shouldn't be a big issue. And for normal cases (no fork) it _would_ be enough to tweak `RollforwardBlock` to process special txes 2cebb48306 (indexes are written directly to disk with no cache so they should be fine already). However `CQuorumBlockProcessor::GetQuorumBlockHash()` relies on `chainActive` which is not yet initialized at that moment, so we also need a (bit hack-ish) tweak to skip some checks to make it work bedf9b93cf.

**Comment by 10xcryptodev** [comment]
@UdjinM6 thanks! I also agree with this temporary solution to get back working the dbcrash test, the dbcrash with fork is really a edge case. Updated with your suggestions.
I'm in the process of understanding evodb technology, but one question that came my mind is: why is necessary save the best block in evodb if txdb already has this information?

**Comment by UdjinM6** [comment]
The best block hash is just a "checkpoint" for these db-s i.e. "we processed block data for <evo/coin> db up to this block". Each db has this checkpoint simply to be able to take care about its own consistency.

File: src/init.cpp [hunk 1]
```diff
@@ -1900,6 +1900,13 @@ bool AppInitMain()
                 // The on-disk coinsdb is now in a good state, create the cache
                 pcoinsTip.reset(new CCoinsViewCache(pcoinscatcher.get()));
 
+
```
Comment (PastaPastaPasta) [inline src/init.cpp:1903]:
```suggestion
```
